internal/pkg/artifact/config.go

```diff
@@ -46,6 +48,38 @@ type Config struct {
 	httpcommon.HTTPTransportSettings `config:",inline" yaml:",inline"` // Note: use anonymous struct for json inline
 }
 
+type Reloader struct {
+	log *logger.Logger
+	cfg *Config
+}
+
+func NewReloader(cfg *Config, log *logger.Logger) *Reloader {
+	return &Reloader{
+		log: log,
+		cfg: cfg,
+	}
+}
+
+func (r *Reloader) Reload(rawConfig *config.Config) error {
+	type c struct {
+		Config *Config `config:"agent.download" yaml:"agent.download" json:"agent.download"`
+	}
+
+	cfg := &c{
+		Config: DefaultConfig(),
+	}
+	if err := rawConfig.Unpack(&cfg); err != nil {
+		return err
+	}
+
+	r.log.Debugf("Source URI changed from %q to %q", r.cfg.SourceURI, cfg.Config.SourceURI)
+	if cfg.Config.SourceURI != "" {
+		r.cfg.SourceURI = cfg.Config.SourceURI
+	}
+
+	return nil
+}
+
```

Comment:
[Blocker]
Sorry but I'm a bit puzzled here. How does it changes the source URI the agent uses? The changes made by the `Reload` method seem to be only local. It only affects `r.cfg.SourceURI` that isn't used anywhere else. Perhaps you forgot to change the `SourceURI` returned by `DefaultConfig` as [elastic-agent/pull/686](https://github.com/elastic/elastic-agent/pull/686/files#diff-9d7e4281df61a79594dd5f0bd75f133f079f3e5932a5909be12c8dca69842ad7R108-R109) did?

Also, it'd be good to add a test for that

Reply:
r has a pointer to a config on a heap. this one is propagated down the stack to each component. 
downloader has the pointer to the same thing so when it tries to resolve URI it will pick it up.
same logic applied for monitoring reloader.